mm2src/coins/utxo/utxo_common.rs

```diff
@@ -3189,6 +3282,192 @@ fn increase_by_percent(num: u64, percent: f64) -> u64 {
     num + (percent.round() as u64)
 }
 
+pub async fn block_header_from_storage_or_rpc<T>(
+    coin: &T,
+    height: u64,
+    storage: &Option<BlockHeaderStorage>,
+    header_params: Option<UtxoBlockHeaderVerificationParams>,
+) -> Result<(BlockHeader, bool), MmError<GetBlockHeaderError>>
+where
+    T: AsRef<UtxoCoinFields> + UtxoCommonOps + MarketCoinOps,
+{
+    let client = match &coin.as_ref().rpc_client {
+        UtxoRpcClientEnum::Native(_) => {
+            return MmError::err(GetBlockHeaderError::NativeNotSupported(
+                "Native client not supported".to_string(),
+            ))
+        },
+        UtxoRpcClientEnum::Electrum(client) => client,
+    };
+    match storage {
+        None => {
+            let bytes = client.blockchain_block_header(height).compat().await?;
+            let header: BlockHeader = deserialize(bytes.0.as_slice())?;
+            Ok((header, false))
+        },
+        Some(storage) => match storage.get_block_header(coin.ticker(), height).await? {
+            None => {
+                let bytes = client.blockchain_block_header(height).compat().await?;
+                let header: BlockHeader = deserialize(bytes.0.as_slice())?;
+                match header_params {
+                    None => Ok((header, false)),
+                    Some(params) => {
+                        let (headers_registry, headers) =
+                            utxo_common::retrieve_last_headers(coin, params.blocks_limit_to_check, Some(height))
+                                .await?;
+                        match spv_validation::helpers_validation::validate_headers(
+                            headers,
+                            params.difficulty_check,
+                            params.constant_difficulty,
+                        ) {
+                            Ok(_) => {
+                                storage
+                                    .add_block_headers_to_storage(coin.ticker(), headers_registry)
+                                    .await?;
+                                Ok((header, true))
+                            },
+                            Err(_) => Ok((header, false)),
+                        }
+                    },
+                }
+            },
+            Some(header) => Ok((header, true)),
+        },
+    }
+}
+
+pub async fn retrieve_last_headers<T>(
+    coin: &T,
+    blocks_limit_to_check: u64,
+    current_block: Option<u64>,
+) -> Result<(HashMap<u64, BlockHeader>, Vec<BlockHeader>), MmError<SPVError>>
+where
+    T: AsRef<UtxoCoinFields> + UtxoCommonOps + MarketCoinOps,
+{
+    let electrum_rpc_client = match &coin.as_ref().rpc_client {
+        UtxoRpcClientEnum::Native(_) => return MmError::err(SPVError::UnknownError),
```

Comment:
I was thinking to not spawn the header loop in native mode since we use only electrum client and therefore maybe use `https://doc.rust-lang.org/std/macro.unreachable.html` because `validate_spv_proof` already return `Ok` before this code is reached, and the header storage loop will not be spawned in native, so this code will indeed be unreacheable - if there is any compile time solution such as <https://dlang.org/spec/version.html> <https://nim-lang.org/docs/manual.html#implementation-specific-pragmas-compileminustime-define-pragmas>

I guess it's possible to have a `cargo` feature `native/electrum` which can therefore be used at compile time - do we already have something like that?

Reply:
I mean a different thing: the `retrieve_last_headers` can be a method of `ElectrumClient`. You don't need it to be generic over the coin, as I can see. Instead of adding `unreachable`, we should try making such pattern matching statements not needed by refactoring.